platform/rct/hal/usart_stm32/usart_device_stm32.c

```diff
@@ -454,3 +476,200 @@ void USART3_IRQHandler(void)
 {
         usart_generic_irq_handler(usart_data + UART_AUX);
 }
+
+static bool dma_rx_isr(volatile struct usart_info *ui)
+{
+        volatile uint8_t* tail = ui->dma_rx.ptr;
+        if (!tail)
+                return false;
+        ui->dma_rx.ptr = NULL; /* This indicates a txfer in progress */
+
+        volatile uint8_t* const buff = ui->dma_rx.buff;
+        volatile uint8_t* const edge = buff + ui->dma_rx.buff_size;
+        volatile uint8_t* const head = edge - (uint16_t) ui->dma_rx.chan->CNDTR;
+        xQueueHandle queue = serial_get_rx_queue(ui->serial);
+        portBASE_TYPE task_awoke = pdFALSE;
+
+        while (tail != head) {
+                uint8_t val = *tail;
+                xQueueSendFromISR(queue, &val, &task_awoke);
+
+                if (++tail >= edge)
+                        tail = buff;
+        }
+
+        ui->dma_rx.ptr = head;
+        return task_awoke;
+}
+
+void DMA1_Channel5_IRQHandler(void)
+{
+        /* Clears all pending IRQs for DMA channel 5 */
+        DMA_ClearITPendingBit(DMA1_IT_GL5);
+
+        volatile struct usart_info *ui = usart_data + UART_WIRELESS;
+        portEND_SWITCHING_ISR(dma_rx_isr(ui));
```

Comment:
so we're both calling the rx_isr handler from both the timer *and* the DMA IRQ handler? Why not just always from the timer?